core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithDerivedTypeVisitor.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.avro;
+
+import java.util.Deque;
+import java.util.List;
+import org.apache.avro.Schema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Avro schema with Type visitor, where Type is guaranteed to be derived from Avro schema, and Avro
+ * schema does not necessarily have field IDs (unlike {@link AvroSchemaWithTypeVisitor}). Avro
+ * schema and Iceberg schema fields are matched by position.
+ */
+public abstract class AvroSchemaWithDerivedTypeVisitor<T> {
```

Comment:
How is this different from the `AvroWithPartnerByStructureVisitor`? Couldn't we use that instead?

Reply:
So I tried `AvroWithPartnerByStructureVisitor `, but could not modify `visitUnion` to call `fieldNameAndType` since the former is static and the latter is not. `fieldNameAndType` is required to access a union option's respective Iceberg type.

Reply:
`fieldNameAndType` is a method on the visitor. In `visitUnion` you should be able to call `visitor.fieldNameAndType`.

Reply:
Moved to `AvroWithPartnerByStructureVisitor`. Introduced intermediate class `AvroWithTypeByStructureVisitor` which could be helpful for other Avro + Iceberg schema structural traversals.

Reply:
Is this visitor needed? The visitor is now based on `AvroWithPartnerByStructureVisitor` so I don't think we need to add this.

Reply:
Good catch. Should have been deleted after the refactor. It is gone now.